Add persistent command history

SlySven opened the pull request:
#### Brief overview of PR changes/additions
Saves each command line's history when a profile is closed and restores the data the next time the profile is opened. Each command line is handled separately.

#### Motivation for adding to Mudlet
It is a bounty item!

#### Other info (issues closed, discussion etc)
This works for command lines of all types, and saves them all separately in the profile's base directory.

This should close https://github.com/Mudlet/Mudlet/issues/2007.

We'll avoid using the names for the extra command lines as part of a filename by instead storing them in a `QSettings` per profile `profile.ini` format file.

Using such a file to store per-profile details is something that has been on my TODO list for years (I prototyped code to do it at least as far back as 2016 I think) - so I have added two functions to the Host class that hopefully can eventually replace the existing `readProfileData(...)` and `writeProfileData(...)`.

Also added code to fix a corner case where a TCommandLine has a name containing '/' or '\'.

Also added a limit to the number of command line history entries to save. Otherwise it will grow indefinitely large as every single entry is retained.  A knob for this has been provided on the profile preferences and it is saved with the profile's XML game save file (not in the base  directory of the profile). It covers the range of 0 to 10,000 entries with a default of 500 and a logarithmic step size (1, 2, 5, 10) for each multiple of 10. This value is applied to ALL command lines in a profile.

Note that this limit is only applied when the commend history is saved, it can still grow to be larger than the limit whilst the profile is active!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Improve: add persistent command history

This works for single word named command lines of all types, and saves them
all separately in the profile's base directory.

This should close #2007.

I am not yet happy with the save format - it does save the history in the
order that the entries are in in the
`(QStringList) TCommandLine::mHistoryList`
member in a UTF-8 form (which makes for easier external modification if
wanted (rather then Qt's default UTF-16 internal `QString`) but the
`QDataStream` format saves a `QByteArray` with a leading int32_t size
element - which, while good from a maintaining integrity point of view
makes it not amenable to external editing.

Testing also needs to be carried out in (possibly on a Window system) to
check that things still work with names that are more challenging for an
OS - e.g. including '\\' or '/' or ':'s...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: handle names for the command line history files better

We'll avoid using the names for the extra command lines as part of a
filename by instead storing them in a per profile `.ini` format file.

Using such a file to store per-profile is something that has been on my
TODO list for years (I prototyped code to do it many years ago) - so I have
added two functions to the Host class that hopefully can replace the
existing `readProfileData(...)` and `writeProfileData(...)`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: fix corner case where a TCommandLine has name containing '/' or '\'

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: remove size prefix in the command line history save files

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add limit to number of command line history entries to save

Otherwise it will grow indefinitely large as every single entry is
retained.

A knob for this has been provided on the profile preferences and it is
saved with the profile's XML game save file (not in the base directory of
the profile. It covers the range of 0 to 500 entries with a default of 50
and a step size of 10. This value is applied to ALL command lines in a
profile.

Note that this limit is only applied when the commend history is saved,
it can still grow to be larger than the limit whilst the profile is active!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Are you across the work Tim has been doing in https://github.com/Mudlet/Mudlet/pull/6733 ?

SlySven commented:
> Are you across the work Tim has been doing in #6733 ?

I've only just looked at it - I started on this unaware that he was working on it, and when I became aware (yesterday) I deliberately avoided viewing it to avoid accusations of plagiarism.  :innocent: 

Now I have looked and I see that he has only considered the main command line (AFAICT) - mine treats each one separately - and can handle issues that would arise from using the names of additional ones in file names. Initial testing shows it seems to work; but I haven't exercised the mechanism to limit the number of entries to save which IS essential as otherwise the list will grow forever (assuming infinite different command are entered...!)

vadi2 commented:
Hmm, Tim's PR has been open for two weeks - and this is not the first time it's happened, you might want to pay attention to linked PRs:

![image](https://user-images.githubusercontent.com/110988/232184160-577a0fe9-c866-4cbb-b848-1709676f1177.png)

That said, the bounty process does allow this:

> Additionally, in case multiple PRs are targeting the same issue, the decision will be made on a case by case basis.

So I'll put it up to the Decision maker patrons to decide when @atari2600tim's PR is ready.

atari2600tim commented:
> So I'll put it up to the Decision maker patrons to decide when @atari2600tim's PR is ready.

No need, I'm not interested anymore.

> I deliberately avoided viewing it to avoid accusations of plagiarism

I wrote it for an open source project.  The central task is reading and writing an array of strings to file.

SlySven committed:
Revise: add per command-line controls for each of them

This adds three tool-buttons shown to the left of each command-line that,
in order:
* sets the command-line's most recently used commands (up to the
per-profile limit) to be saved in a file.
* sets the next command entered NOT to be saved in the history at all, as
a side effect of the way it works if the exact command has been used before
then entering it again with this option selected WILL erase that particular
entry from the stored history (and when the profile is closed) from the
file backing it up as well. After the command has been entered this option
is automatically deselected and the selected option reverts to the first
one.
* sets the command history for this command line to NOT be saved at all -
as Mudlet would behave before this PR came about. The commands are still
remembered for the current session but they are not written out to a file
though if they have been in a previous session than those will remain in
the file as it will not be updated when the profile is closed.

Also, if the number of lines to be saved is set to zero (the "None") value
in the settings then all three buttons are disabled (greyed out) in all the
command lines for that profile.

To add buttons to control this feature in the `QPlainTextEdit` proved to be
quite tricky. Unlike the `QLineEdit` class there is no provision to add
`QAction`s to either end. Instead I had to create a `TCommandLineWidget`
class to encapsulate a `TCommandLine` as well as the three buttons in a
`QFrame` container widget. Even this was not trivial but was possible to
do - unlike an initial failed attempt that tried to extend the
`TCommandLine` class itself to inherit from those other widgets as well.
The existing methods/interactions with the original `TCommandLine` had to
be revised to either work on the new wrapper class or to access the wrapped
one via that instead of working directly with it.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven committed:
Revise: fixup out-of-order QGridLayout in MODIFIED .ui files

This PR only touches the `./src/ui/profile_preferences.ui` that has such an
out-of-order `QGridLayout` items in the one called `"gridLayout"` which is
the layout for the `QGroupBox` called `"groupBox_accessibility"` - were I
to rename that layout from the generic (probably Qt Designer created) name
I would call it `"gridLayout_groupBox_accessibility"` but I will refrain
from doing that here as it will only cloud the `diff`.

The following `QGridLayout`s that still need attention are not modified by
this PR so will not be "fixed" here:
* `./src/ui/connection_profiles.ui`	--> `"gridLayout"` - out of order
* `./src/ui/custom_lines.ui`	-->	`"gridLayout_3"` - redundant
* `./src/ui/custom_lines.ui`	--> `"gridLayout_style"` - out-of-order
* `./src/ui/custom_lines.ui`	--> `"gridLayout_4"` - redundant
* `./src/ui/delete_profile_confirmation.ui`
                                         	--> `"gridLayout"` - out-of-order
* `./src/ui/vars_main_area.ui`
                    	--> `"gridLayout_frame_vars_main_area"` - out-of-order

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven committed:
Standarize UI files

demonnic commented:
The latest windows test version does not show the title bar and accompanying buttons in windows. Which is to say I cannot resize it, maximize, or easily close it.

I'm also not sure I'm a fan of the 3 new buttons attached to every single command line, it drastically changes the look of anything already using a command line without anything I can do about it.  This also makes it much more unwieldy to try and incorporate it into new geyser objects, for instance a number spinner I'm working on where I'd planned to use a small command line as a number entry option. But with these extra buttons tacked on suddenly this widget is required to be twice as wide just to fit the command line in.

SlySven commented:
> The latest windows test version does not show the title bar and accompanying buttons in windows. Which is to say I cannot resize it, maximize, or easily close it.

Not sure that this is the cause of *that* - TBH it sounds like you have gotten into full-screen mode - which I know some OSes do allow outside of the application itself...

>I'm also not sure I'm a fan of the 3 new buttons attached to every single command line, it drastically changes the look of anything already using a command line without anything I can do about it. This also makes it much more unwieldy to try and incorporate it into new geyser objects, for instance a number spinner I'm working on where I'd planned to use a small command line as a number entry option. But with these extra buttons tacked on suddenly this widget is required to be twice as wide just to fit the command line in.

Yeah, I get where you are coming from - I am thinking that I might be able to get the three options onto the context menu - and perhaps leave a single (`QToolButton`) button showing an icon representing the current state that can be clicked on to change to the next one. Would a single button be too much for you and Geyser?

demonnic commented:
It'd be a lot less difficult in the layout than 3 of them, but ideally it would be something I could hide even still, so I could get the current behavior of "just the input box" again.

SlySven commented:
TBH I now see that it can be done entirely within the context menu of each command line - but I do wonder about the feature's visibility - in the "Forget this line" mode it only stays active for, guess what, the one line before dropping back to the "Save commands" mode and I'm not sure that that will be apparent without some visible status indication, yet finding a way to show that is tricky - it was a major hassle trying to get those buttons incorporated into the `TCommandLine` class and as you can see by the current code as of https://github.com/Mudlet/Mudlet/pull/6767/commits/c77060e9ab17662742973a46c806df277e6da2f1 I could not do that and instead had to put a `TCommandLineWidget` "wrapper" around it - which messes with the API in a significant way. Going to a context menu route (as I am currently working on) is considerably cleaner route but is much less visible...

SlySven committed:
Revise: remove the tool-buttons and move the UI to the context menu

This reduces the amount of disruption this PR was causing to the
`TCommandLine` class. I do have concerns about indicating to the end-user
that either the next or all commands are not being saved for the command
line going forward when we announce the feature that they will be saved
(by default) but I have not yet come up with a foolproof scheme to show
that yet.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven committed:
Tweak: some small adjustments/tweaks needed

Signed-off-by: Stephen <slysven@virginmedia.com>

SlySven committed:
Fix: remove resource file entries for removed images

:man_facepalming: I forgot to do it!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
In use now, with the options moved to the context menu for each command line:
![image](https://user-images.githubusercontent.com/6163092/233807625-383e15f6-e32d-4967-ae3b-67ed601604b5.png)

Note that setting the preference to `None` will disable the options (with additional text to describe the effect/reason):
![image](https://user-images.githubusercontent.com/6163092/233807697-4ae7b507-aae0-4d6c-8174-9cd58c3fb110.png)

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

demonnic commented:
Is it intended that you lose the per-session command history if it's set not to save the commands? That was unexpected for me, as I thought these would only effect if it was persisted across sessions/saved to disk. I imagine most people will leave it to the new default of saving to disk and not necessarily notice a difference, but it was still a little jarring for me.

SlySven committed:
Fix: correct an error made during prior merge and update some comments

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> Is it intended that you lose the per-session command history if it's set not to save the commands? That was unexpected for me, as I thought these would only effect if it was persisted across sessions/saved to disk. I imagine most people will leave it to the new default of saving to disk and not necessarily notice a difference, but it was still a little jarring for me.

No, it isn't :slightly_frowning_face: looking at the code now I think I miscoded something in `(void) TCommandLine::enterCommand(QKeyEvent*)` just trying to fix that...

SlySven committed:
Fix: allow command history to continue to work even if not saved

I had unintentionally disabled the whole command line history system for a
particular `TCommandLine` instance when it was set to not remember its
history. I think this was due to a different initial assumption about what
would be needed that I forgot to revise when I worked things out!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven commented:
@demonnic It should work as intended now...

demonnic reviewed (approved):
functionality seems good now

vadi2 commented:
One thing that's not entirely clear to me is how many commands does the current command history remember during the session max?

vadi2 commented on src/Host.cpp:
```diff
@@ -820,6 +820,13 @@ std::tuple<bool, QString, QString> Host::saveProfile(const QString& saveFolder,
 
+    if (saveFolder.isEmpty() && saveName.isEmpty()) {
+        // This is likely to be the save as the profile is closed
+        qDebug().noquote().nospace() << "Host::saveProfile(...) INFO - called with no saveFolder or saveName arguments "
```
Is this still needed now that this feature is ready? I would prefer the unix "quiet unless there's a problem" approach here.

SlySven replied:
TBH When I coded this I was not entirely - and I am ***still not*** - clear whether this only happens when the profile is closed. If any reviewer has any more information about whether this is true or not I'd like to know.

vadi2 replied:
Hang on... why is this exception even here? It should run whenever profile saving is done, as saving a profile means everything is to be saved!

SlySven replied:
Do we write out all the individual, single item files in the profile's home directory every time the profile is saved? No, we save them when the data is changed. This feature is about saving the profile's command history between sessions which it does.

Does saving it every autosave interval make sense? :man_shrugging:

vadi2 replied:
I think so - Mudlet can crash (unfortunately), or the OS can crash, so why lose this stuff?

vadi2 commented on src/Host.cpp:
```diff
@@ -2144,6 +2151,116 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
+// mHistoryData:
+bool Host::writeProfileIniData(const QString& item, const QString& what)
+{
+    QSettings settings(mudlet::getMudletPath(mudlet::profileDataItemPath, getName(), qsl("profile.ini")), QSettings::IniFormat);
```
To my understanding, the settings object is intended to persist throughout the application's lifetime, as is currently being used - I'm curious as to why it is being initialized each time a value is written here.

SlySven replied:
Nope AFAICT `QSetting`s can be created and destroyed as needed. Destruction just helps to ensure things are synced.

From the Qt Documentation:
> ### Accessing Settings from Multiple Threads or Processes Simultaneously
> QSettings is **reentrant**. This means that you can use distinct QSettings object in different threads simultaneously. This guarantee stands even when the QSettings objects refer to the same files on disk (or to the same entries in the system registry). If a setting is modified through one QSettings object, the change will immediately be visible in any other QSettings objects that operate on the same location and that live in the same process.
>
> QSettings can safely be used from different processes (which can be different instances of your application running at the same time or different applications altogether) to read and write to the same system locations, provided certain conditions are met. For `QSettings::IniFormat`, it uses advisory file locking and a smart merging algorithm to ensure data integrity. The condition for that to work is that the writeable configuration file must be a regular file and must reside in a directory that the current user can create new, temporary files in. If that is not the case, then one must use `setAtomicSyncRequired()` to turn the safety off.

vadi2 replied:
Okay. I think it would be better design if the object persisted throughout the lifetime of the Host, having Qt do "advisory file locking and a smart merging algorithm to ensure data integrity" every time we set a value does not sound clean.

SlySven replied:
That second paragraph bit is all about different applications isn't it - that doesn't seem relevant here, but is useful to know if we ever go off down the "single profile per application instance" path?

vadi2 replied:
I checked in with the Qt community and this works fine for other folks, so... let's give it a go!

vadi2 commented on src/Host.cpp:
```diff
@@ -2144,6 +2151,116 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
+}
+
+// Returns the name for the file - either "command_history_main" for the main
+// command line or "command_history_##" (where ## is a leading zero number,
```
What does the number mean? The comment does not explain. I'm wondering because command lines are given names, not numbers.

How about this comment instead, what do think?

```
// Retrieves the command line history settings for the given command line type and name.
// For the main command line, reads the saveCommands and forgetNextCommand settings from the profile.ini.
// For a sub-command line, first tries to read the settings from the profile.ini. If no settings exist yet,  
// generates a new file name and saves default saveCommands and forgetNextCommand settings to the profile.ini.
// Returns the file name, saveCommands setting, and forgetNextCommand setting for the given command line.
```

SlySven replied:
Each time a new command line history file is needed for a differently named command line the `UsedIndexes` number is increment and used for that command line, and the mapping for the command line name to the file name stored in the `QSetting`'s `profile.ini` file - doing it in this manner is to avoid nasty problems when the name used by the user is not one that can be used in a file name.
```cpp
// Retrieves the command line history settings for the given command line type
// and name, for any command line this reads the saveCommands and
// forgetNextCommand settings from the profile.ini that is intended to
// eventually replace all the other single data item files in the profile's
// home directory. For the main command line the file name to use is predefined
// but for any other command line, the mapping from a number suffixed file to a
// command line name is first looked for; if not found (so no settings exist
// yet) a new file name and default saveCommands and forgetNextCommand settings
// are created and also stored before being returned.
```

vadi2 replied:
How about this version with increased text clarity?

> This function retrieves command line history settings based on the given command line type and name. It reads the saveCommands and forgetNextCommand settings from the profile.ini file, which is intended to replace all other single data item files in the profile's home directory. For the main command line, a predefined file name is used, while for any other command line, the function first looks for a mapping from a number-suffixed file to a command line name. If the mapping is not found, a new file name and default settings for saveCommands and forgetNextCommand are created, stored, and returned.

SlySven replied:
:+1:  Though I'd change it from `function` to `method` - this *is* C++ don'tcha' know! :laughing:

vadi2 commented on src/Host.cpp:
```diff
@@ -2144,6 +2151,116 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
     return QString();
 }
 
+// This pair might eventually replace the ones without the "Ini" but for now
```
I think a developer seeing this comment for the first time would struggle to understand what "the ones" are, perhaps clarify that it's talking about `writeProfileData`?

SlySven replied:
Done in a1d1110 ...

vadi2 commented on src/Host.cpp:
```diff
@@ -2144,6 +2151,116 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
+    }
+    QString localName{name};
+    localName.replace(QRegularExpression(qsl("[\\/]")), qsl("_"));
+    // We use a '/' in the name as that donotes a grouping (section) within the
```
Codespell seems to be kaput :D

SlySven replied:
Agreed, fixed in a1d1110.

vadi2 commented on src/Host.cpp:
```diff
@@ -2144,6 +2151,116 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
+// Because '/' and '\\' are used by the QSettings class in key names for
+// special purposes we MUST filter them out in the name, we'll replace them
+// with '_'s:
+std::tuple<QString, bool, bool> Host::getCommandLineHistorySettings(const TCommandLine::CommandLineType type, const QString& name)
```
getCommandLineHistorySettings 30 character is a mouthful 😅 how about just getCmdLineSettings which happen to only store history for now?

SlySven replied:
Okay, changed to that in a1d1110.

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -597,6 +597,44 @@
+             <number>0</number>
+            </property>
+            <property name="maximum">
+             <number>500</number>
```
I wouldn't limit to 500, seems arbitrary, if someone wants to store a few thousand they should be able to - consumers computers today have up to TB of storage

SlySven replied:
How far back in your command line history do you go when you are MUDding? Remember each time you reuse a command it gets pulled to the front of the history again. To me this seems a reasonable number of "frequently reused commands" and do you really want a knob that has a range of, say 0 to 10,000 (with a current step of 10 - so with 1,001 notches/steps on the adjustment).

vadi2 replied:
One can always type a number in, pressing on the little buttons is optional (and is the least quickest way to go about it!)

demonnic replied:
I feel like 500 is too lower, and would like to point out that you're not always going back through your history one item at a time, sometimes you're typing the first part of a command in and then jumping up in the history to the first thing which matches the part you typed in. Having more than 500 lines to pull from between sessions could prove useful.

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -597,6 +597,44 @@
+             <number>10</number>
+            </property>
+            <property name="value">
+             <number>50</number>
```
If I learnt one thing is that computers have gotten really powerful in the last decade Mudlet was around, would feel better with a 1000 default

SlySven replied:
As per previous comment really, I just want a set of sensible values for three things:
* maximum number, I said `500` you say "thousands"
* default number, I said `50` you say `500`
* step-size on the control, I said `10`,  you haven't - but if you keep my value and apply it to, say `5,000` that means `51` steps from minimum (`none` - disabled no storage for ANY command line) for my current values to `5,001` for that interpretation I am placing on your suggestions. TBH I would probably change it to `50` or `100` which would scale it back a bit.

:thinking:  Assuming that one did save, say `5,000` commands, how long would it take to scroll back through them for a freshly loaded profile and would one really want to have to decide whether to save `5,000` or only `4,990` ?

vadi2 replied:
I don't this the step size to be vital, most people select the field and type a number in.

demonnic replied:
I also feel like 50 for a default is kind of low. I can burn through 50 commands pretty quickly, and don't know that only the last 50 commands from any given session would be that beneficial to carry over per se. While I understand a desire to set sensible defaults I'm not sure 50 and 500 are actually that sensible in 2023. 

I also agree that step size is not super important, but why not just increase it by the same factor and go for default 500, max 5000, and step of 100? Though I might set the max in particular even higher, to say 10k or more. It's just a plain text file, yeah? As long as it doesn't reload itself in its entirety every time you write a new command to the history or anything it should be fine.

vadi2 commented on src/TCommandLine.h:
```diff
@@ -127,9 +132,34 @@ public slots:
 Q_DECLARE_OPERATORS_FOR_FLAGS(TCommandLine::CommandLineType)
 
+#if !defined(QT_NO_DEBUG)
+inline QDebug& operator<<(QDebug& debug, const TCommandLine::CommandLineType& type)
```
👍

SlySven replied:
I needed it during debugging and these sort of `QDebug()` helpers can improve the debug output - I just haven't had the nerve to do a `TRoom` one yet... :laughing:

vadi2 replied:
https://github.com/Mudlet/Mudlet/pull/6814

vadi2 commented on src/TCommandLine.cpp:
```diff
@@ -693,6 +703,69 @@ void TCommandLine::mousePressEvent(QMouseEvent* event)
             QAction* action_removeWord = nullptr;
             QAction* action_dictionarySeparatorLine = nullptr;
+
+            QActionGroup* actionGroup_commandHistory = new QActionGroup(popup);
```
Similar to a comment I left in https://github.com/Mudlet/Mudlet/pull/6743#pullrequestreview-1412819376, this will take up a lot of space in the right-click menu while it will be needed very rarely, and that doesn't balance.

This functionality to ignore things is necessary, I agree with that - how about it's available via a Lua API instead? Then those who want to use it can create their own right-click menus, buttons, or whatever else.

It can be a simple toggle per-command line whenether command line history should be saved or not.

SlySven replied:
This adds 3 lines plus a separator to the context menu - but the spelling can add ~~loads and loads and  loads~~ ***lots*** more. If you are typing an ultra secret password and want to be certain it isn't remembered do you want to load in an additional Lua module that might do something sneaky instead (but makes the context menu 1 more line plus a separator normally or 2 lines plus 2 separators when the module is used) or  just right click on the context menu and select an option that is known to be hard coded into the application and always visible there and becomes familiar with repeated use.

In my mind your UI/UX concept does not sound *that* compelling...

vadi2 replied:
> but the spelling can add loads lots more. 

Yeah, and having tried it I think it is too verbose (adds 5 lines even when there's nothing). I don't want to go down this route here too. Way too few people will have a need for this, in too little scenarios, so how much UI space it'll take up is not in balance.

SlySven replied:
Ah,  I've just thought of something about the UI I did that wouldn't fly so well with splitting the functionality up into two separate things:

When you select the "forget the next line only" option it stays in that state until a command is entered into that command line (even between sessions) *then* once a command has been entered it jumps back to the "Save command history" option - so it gives a visible (post-usage) indication that it has taken place.

If the "don't remember option" is selected then "forget the next line" is largely pointless (not entirely because if it is selected from either of the other states then any instances of the command then given is nuked from the current session command history - and if that history *is* saved, from that as well).  By doing this whilst "disconnected" it also provides a means to remove a command (say an entered password) from the history file should one have accidentally have gotten saved without having to track down the history file and edit it by hand - which is possible BTW.

Anyhow, I don't think it would be easy to separate out the two different features and leave one in the context menu and the other down to a Lua package and provide the (relatively simple):
* "save command history"
* "don't save the next command" - changing back to "save command history" after next command
* "don't save command history"

UI.

demonnic replied:
The right click menu is getting a little bit big, but I still prefer that to the buttons which were introduced previously.
I think it would be a good thing to be able to set whether a command line saves its output from the Lua API regardless of whether it remains in the right click menu by default though. So you can set the value as a Geyser constraint as well when creating one.

vadi2 commented on src/TCommandLine.cpp:
```diff
@@ -1412,3 +1504,111 @@ void TCommandLine::slot_adjustAccessibleNames()
+    }
+
+    QString pathFileName{mudlet::self()->mudlet::getMudletPath(mudlet::profileDataItemPath, pHost->getName(), mBackingFileName)};
+    QSaveFile historyFile(pathFileName, this);
```
+1

vadi2 reviewed (commented):
Would be great to finally have this :)

SlySven commented:
> One thing that's not entirely clear to me is how many commands does the current command history remember during the session max?

As it is before this PR - unlimited (well there is a theoretical `QStringList` maximum but I do not see how anyone could reach it before Mudlet explodes in some other way...!)

SlySven commented on src/Host.cpp:
```diff
@@ -2144,6 +2151,116 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
     return QString();
 }
 
+// This pair might eventually replace the ones without the "Ini" but for now
+// are just used to store some information about one or more TCommandLine's
+// mHistoryData:
```
```suggestion
// writeProfileIniData(...) and readProfileIniData(...) might eventually
// replace writeProfileData(...) and readProfileData(...) but for now are just
// used to store some information about one or more TCommandLine's mHistoryData:
```

SlySven replied:
Done in a1d1110.

SlySven committed:
Revise: some changes from peer-review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:bulb: Ah ha! I think I can see a way to improve the context menu situation (and not have any more lines than you wanted @vadi2 . Let me see if I can code it AND a Lua getter/setter pair. AND a non-linear number of lines selector for the preferences..

SlySven committed:
Revise: change a comment as suggested in peer-review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: move command history options to a sub-menu

So that they do not take up so much space on the main context menu for the
command-line.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven committed:
Revise: change the range of command line history entries that can be saved

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
@demonnic Do you want Lua functions to:
* get/set the per-command-line "save all lines/don't save next line/don't save any lines" options
* get/set the per-profile "size of command line history to be saved" option
* or do you want all of the above?

demonnic commented:
> @demonnic Do you want Lua functions to:
> 
> * get/set the per-command-line "save all lines/don't save next line/don't save any lines" options
> * get/set the per-profile "size of command line history to be saved" option
> * or do you want all of the above?

I mean, ideally all of the above, but the first one would be the more important of the two if I had to choose.

SlySven committed:
Improve: add Lua functions to get and set per-command line history options

Adds:
* `getCommandHistoryOptions([commandLine]) - returns a boolean which is
indicates if the commands entered on the command line are to be saved and
if not, why. A second return value gives more details. The combinations
are:
  * `true`, `"enabled (%1 lines will be saved)"` - where %1 is the number
set globally in the per-profile preferences
  * `false`, `"disabled by profile global preference"`
  * `false`, `"disabled for all commands"`
  * `false`, `"disabled for next command only"`
* `setCommandHistoryOptions([commandLine,] forgetNextCommand
                                                  [, forget all commands])`

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: using `QComboBox` to set count of number of commands to save

I forgot to specify the dateRole to find the data for given that I was also
specifying the matching flags.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Improve: insert commandLineHistorySaveSize option to Lua (g|s)etConfig(...)

This is so that the profile wide command line setting from the preferences
is also available to the Lua sub-system.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

SlySven commented on src/XMLimport.cpp:
```diff
@@ -907,6 +907,13 @@ void XMLimport::readHost(Host* pHost)
+    } else {
+        // This is the default valuem though prior to the introduction of this
+        // it would have effectively been zero:
+        pHost->setCommandLineHistorySaveSize(50);
```
Oh, need to change the original default of 50 to match the 500 in the preferences now...

SlySven committed:
Revise: align default value for the command line save size

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
How about this take to balance out what I mentioned in https://github.com/Mudlet/Mudlet/pull/6767#discussion_r1186251173 ? This would be **in addition** to the right-click `Command history options...` menu you've already added. 

![Selection_010](https://github.com/Mudlet/Mudlet/assets/110988/92e89ddb-1e89-44f3-b5e4-e71bd1265e31)

That way most folks won't have to deal with the option and it won't get in their way, and the few that do can have it for their needs.

Remember that Mudlet stored character passwords in plaintext for years until secure storage came about, but hardly anyone was bothered by it. What engineers think is important can sometimes be different from what is actually important to the players :D

(I have a bit more feedback on the menu, but want to take it one step at a time)

SlySven commented:
"Show fine-grained command history options" - what do you mean? Which command line - all of them or just the "main" one? If you want to simply turn the whole option off I could arrange for the context menu option to be *hidden* rather than just "greyed-out" and disabled for all command-lines if you select "None" in the preferences but I was thinking that showing that there was an option but it had been switched off/overridden by a master prefernce was a more informative way to go.

I think that having separate controls for each command line is required given that we keep separate command histories for each command line during the session.

vadi2 commented:
> Which command line - all of them or just the "main" one?

This would apply to all of them.

> If you want to simply turn the whole option off 

I want the history persistence to stay, but without the additional right-click menus, because I think they would be shown far more than they are actually needed or desired to be seen.

> I think that having separate controls for each command line is required given that we keep separate command histories for each command line during the session.

This doesn't apply to controls, but rather visibility of them. If someone wants to see them, they can enable them, but I don't see the larger majority of players needing to see them - and they'll get in the way.

SlySven commented:
> I want the history persistence to stay, but without the additional right-click menus, because I think they would be shown far more than they are actually needed or desired to be seen.

Look I have taken the controls to a sub-menu so that the current option is visible in the main menu (**and shows the status of it for *that* command line**) but removing that altogether seems unhelpful. Remember that the "forget next command" status DOES change after a command has been entered (back to the "Save commands" one) and the icon changing is the simplest way to show that has happened - or is pending to happen.

> This doesn't apply to controls, but rather visibility of them. If someone wants to see them, they can enable them, but I don't see the larger majority of players needing to see them - and they'll get in the way.

Given that the setting can be manipulated by the Lua sub-system now - how can the end-user determine what the current state is? I suppose it might be possible to reproduce the UI as I have it now with an additional Lua package but I will be perfectly frank and say I am not the person to do it like that - and I am not sure the visual UI (with the automatically greyed out icons and the tooltext changes should the system be disabled so as to behave like Mudlet does without this PR) could be reproduced easily anyhow.

As for "getting in the way" I don't see how this is any more intrusive than anything else we have put in the context menu? I can't recall the last time I used the "Search" feature but there isn't a way to hide that currently, yet I don't mind that being there...

:thinking: Has there been any feedback on this PR in either direction from any testers yet?

vadi2 commented:
There is no search feature in the right-click menu - 

![image](https://github.com/Mudlet/Mudlet/assets/110988/75077b55-6772-427f-8a65-dc47d4680c3a)

My take on this is still the same as in https://github.com/Mudlet/Mudlet/pull/6767#discussion_r1186251173: nothing about command persistent should be added to the right-click menu default and out of the box as too few people in too few situations will find it useful. I feel strongly about this as I believe it is important for Mudlet to remain modern. This is learning from mistakes of other clients that have added so many niche options that players consider them complex and difficult to use. 

I would like to see if I'm alone in this or no, so anyone following this, please weigh in!

Delwing commented:
I think this is not very frequent thing to change. More like set once and never change it again. Coukd be safely in regular settings screen window only.

demonnic commented:
I don't foresee needing to see these settings that often myself, and I'm what you might call a "power user" in that regard. This context menu is already getting pretty lengthy, and I wouldn't be against finding a way to trim/condense it even further, but that should perhaps be a separate ticket to discuss actions beyond adding (or rather not adding) more to the context menu as part of this PR.

I am hoping to get the time to test the new Lua API functions later today and weigh in on that aspect as well, haven't had quite as much time over the last week as I'd hoped I would.

keneanung commented:
My thoughts on this:
1. Have the amount of saved lines in the settings dialog
2. Don't put an upper limit the amount of saved lines. If need be, add a warning hint somewhere. But I don't think it will be necessary.
3. Apply one setting to all input elements.
4. Have the "don't keep this command line's history" setable via lua API
5. Don't keep the command line history as long as IAC DONT ECHO (should catch passwords) is active and make it available via lua (for those that really want/need it). I can't see many people remembering a right-click menu option before entering sensitive data. It's either in hindsight or just doesn't matter. (Maybe add an API function to remove certain items?)

SlySven commented:
> There is no search feature in the right-click menu

Sorry, I was mixing it up with the `TConsole` context menu.

> I don't foresee needing to see these settings that often myself...

As for never changing the option I agree that once you have decided between the "save"/"don't" save options you are not going to change between them - however if you go for the "save" option you may will want access to the "don't save the next one" option occasionally and accessing that from the context menu still seems like the most intuitive method. 

@keneanung 's point about password entry has relevance but directly coding that to the server negotiating to perform ECHOing is not a good idea - whilst we already prevent reproducing the entered command on the output (`TMainConsole`) window in that situation, doing more than that automatically seems unwise because the user can be doing something locally and want to see what they are typing (for instanced: something using the `lua` alias).

I do hope that in the future the "forget the next command" can be combined with a context menu option (which the user controls directly or has the option to couple ***or not*** to the Server ECHOing) to switch to a password entry mode. 

However that is not something that Qt offers to us because the widget from which we derive the `TCommandLine` class from does not have that (though others do); so we will have to come up with a method/process to do that ourselves (perhaps intercepting each keypress and replacing each character entered with a "*" after a fraction of a second, though left/right arrows and delete/backspace keypresses will need special handling!)

keneanung commented:
> whilst we already prevent reproducing the entered command on the output (TMainConsole) window in that situation, doing more than that automatically seems unwise because the user can be doing something locally and want to see what they are typing (for instanced: something using the lua alias).

I don't want to hide more from the consoles. I just don't want to save command line entries as long as that option is on.

vadi2 commented:
We're all in agreement then, it's desired that the command line's UI does not change in any way (nothing on the right-click menu nor as attached buttons), and I agree with @keneanung spelling out how it should work.

This is a great feature, just the current UI aspects of it are a bit much.

SlySven commented:
I don't think you all are getting one feature of this enhancement that will turn into a defect if there is no easy way to control it!

If you are saving the command history, which we agree, most people will want on by default, what steps do we expect users to take should they want to type in something secret that they do not want written into a plain text file?

They must:
* either create a **separate** command line which permanently is set to NOT save a command history (which will be doable via lua functions only, if we do not provide a UI means) - the secret will be kept in the history for that command line (and can be recalled within that session but not remember beyond that)
* or they need to create a **separate** command line so that *one* of them can call the function that causes the ***other*** one to forget the *next* command entered into it (not remembered at all, even within the session) - remember that you cannot use the function from the command line that you want it to apply to, because the call of that function is the command that gets "forgotten".

vadi2 commented:
It could be a problem but I can't judge how realistic it will be in practice, let's see if the current solution:

> either create a separate command line which permanently is set to NOT save a command history (which will be doable via lua functions only, if we do not provide a UI means) - the secret will be kept in the history for that command line (and can be recalled within that session but not remember beyond that)

Works and if it doesn't, we'll improve on it. And in the meantime, there will be no space taken up in the UI for this uncommon scenario (which everyone seems to agree is a good idea not to do).

SlySven commented:
Well, if there is no UI and you really want just the:
> create a separate command line which permanently is set to NOT save a command history (which will be doable via lua functions only, if we do not provide a UI means) - the secret will be kept in the history for that command line (and can be recalled within that session but not remember beyond that)

then I can then rip out a whole wodge of code associated with the "forget the next command (only)" functionality... :roll_eyes:

SlySven committed:
Merge branch 'development' into Improve_addPersistantCommandHistory

Conflicts resolved in:
* src/TCommandLine.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: cripple functionality as requested

Remove the UI and all the code associated with the "forget next command"
option.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/TCommandLine.cpp:
```diff
@@ -885,41 +892,41 @@ void TCommandLine::mousePressEvent(QMouseEvent* event)
-    QPlainTextEdit::mousePressEvent(event);
+    // Process any other possible mouseReleaseEvent - which is default context
+    // menu handling - and which accepts the event:
+    QPlainTextEdit::mouseReleaseEvent(event);
```
Nice catch!

vadi2 reviewed (approved):
🚀
